File: backend/biz/schemas/platform_overview.py
"""平台首页总览接口模型。"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from biz.schemas.common import ApiResponse

AssetOverviewKey = Literal["entity", "relation", "property"]


def _to_camel(value: str) -> str:
    first, *rest = value.split("_")
    return first + "".join(part.capitalize() for part in rest)


class CamelCaseModel(BaseModel):
    """Python 使用 snake_case，接口响应序列化为前端习惯的 camelCase。"""

    model_config = ConfigDict(alias_generator=_to_camel, populate_by_name=True)


class AssetOverviewGroup(CamelCaseModel):
    key: AssetOverviewKey
    title: str
    total: str
    total_label: str
    added: str
    added_label: str


class AssetChangeRow(CamelCaseModel):
    type: str
    object: str
    change: str
    source: str
    time: str


class LatestChange(CamelCaseModel):
    time: str
    type: str
    domain: str
    title: str
    detail: str
    impact: str
    to: str


class ManagementRisk(CamelCaseModel):
    title: str
    detail: str
    detail_to: str
    review_to: str


class StructureItem(CamelCaseModel):
    label: str
    schema_name: str = Field(alias="schema")
    count: str
    ratio: int
    tone: str


class PlatformOverviewData(CamelCaseModel):
    platform_status: str
    pending_batch_count: int
    updated_at: str
    asset_overview_groups: list[AssetOverviewGroup]
    asset_change_rows: dict[AssetOverviewKey, list[AssetChangeRow]]
    latest_changes: list[LatestChange]
    management_risks: list[ManagementRisk]
    entity_structure: list[StructureItem]
    relation_structure: list[StructureItem]
    data_mode: Literal["live", "partial", "mock"] = "mock"
    data_sources: dict[str, str] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)


class PlatformAssetSummaryData(CamelCaseModel):
    platform_status: str
    pending_batch_count: int
    updated_at: str
    data_mode: Literal["live", "partial", "mock"]
    data_sources: dict[str, str]
    warnings: list[str]
    items: list[AssetOverviewGroup]


class PlatformAssetChangesData(CamelCaseModel):
    asset_type: AssetOverviewKey
    rows: list[AssetChangeRow]
    data_source: str


class PlatformActivityData(CamelCaseModel):
    items: list[LatestChange]
    data_source: str


class PlatformRiskData(CamelCaseModel):
    items: list[ManagementRisk]
    data_source: str


class PlatformStructureData(CamelCaseModel):
    entity: list[StructureItem]
    relation: list[StructureItem]
    data_source: str


class PlatformOverviewResponse(ApiResponse):
    data: PlatformOverviewData


class PlatformAssetSummaryResponse(ApiResponse):
    data: PlatformAssetSummaryData


class PlatformAssetChangesResponse(ApiResponse):
    data: PlatformAssetChangesData


class PlatformActivityResponse(ApiResponse):
    data: PlatformActivityData


class PlatformRiskResponse(ApiResponse):
    data: PlatformRiskData


class PlatformStructureResponse(ApiResponse):
    data: PlatformStructureData

File: frontend/src/api/platformOverview.ts
import { http } from './http'
import {
  unwrapApiResponse,
  type ApiResponse,
} from './graphSearch'

export type AssetOverviewKey = 'entity' | 'relation' | 'property'

export interface AssetOverviewGroup {
  key: AssetOverviewKey
  title: string
  total: string
  totalLabel: string
  added: string
  addedLabel: string
}

export interface AssetChangeRow {
  type: string
  object: string
  change: string
  source: string
  time: string
}

export interface LatestChange {
  time: string
  type: string
  domain: string
  title: string
  detail: string
  impact: string
  to: string
}

export interface ManagementRisk {
  title: string
  detail: string
  detailTo: string
  reviewTo: string
}

export interface StructureMember {
  name: string
  count: number
}

export interface StructureItem {
  label: string
  schema: string
  /** 桶内全部非零成员（按计数降序），悬停中文标签的浮窗展示；降级演示数据无此字段 */
  members?: StructureMember[]
  count: string
  ratio: number
  tone: string
}

export interface PlatformOverviewData {
  platformStatus: string
  pendingBatchCount: number
  updatedAt: string
  assetOverviewGroups: AssetOverviewGroup[]
  assetChangeRows: Record<AssetOverviewKey, AssetChangeRow[]>
  latestChanges: LatestChange[]
  managementRisks: ManagementRisk[]
  entityStructure: StructureItem[]
  relationStructure: StructureItem[]
  /** 环形图中心数 = 各分段之和（Σ标签/Σ边类型计数），与分段自洽；资产卡 total 仍是去重口径 */
  entityStructureTotal?: string
  relationStructureTotal?: string
  dataMode: 'live' | 'partial' | 'mock'
  dataSources: Record<string, string>
  warnings: string[]
}

const PLATFORM_OVERVIEW_ENDPOINT = '/v1/platform/overview'

/** 图资产统计随 space（全局图空间选择器当前空间）查询；缺省由后端回落 env 默认空间。 */
export async function getPlatformOverview(space?: string): Promise<PlatformOverviewData> {
  const response = await http.get<
    ApiResponse<PlatformOverviewData>,
    ApiResponse<PlatformOverviewData>
  >(PLATFORM_OVERVIEW_ENDPOINT, { params: space ? { space } : undefined })

  return unwrapApiResponse(response)
}
